From 36eedcbfdb49d1c0f325cdfeeeb1716a6e18b2ae Mon Sep 17 00:00:00 2001
From: Bhatta Sundeep <sbhatta@marvell.com>
Date: Thu, 6 Dec 2018 16:41:08 +0530
Subject: [PATCH 0848/1051] net: octeontx2: initialize fields properly before
 timecounter_init

ptp_ptr->nic should be set before calling timecounter_init.
Otherwise ptp_ptr->nic will dereferenced in ptp_cc_read before it is
set which results in NULL pointer dereference.
Though there is a check whether ptp_ptr->nic is non zero before
dereferencing it invalid pointer dereference happens sometimes
because ptp_ptr->nic may contain garbage since memory for ptp_ptr was
allocated using kmalloc instead of kzalloc. The kernel dump when the
problem occurs is as below:

[ 114.152846] Unable to handle kernel NULL pointer dereference at virtual
address 00000008

[ 115.035950] [<ffff000008a704fc>] _raw_spin_lock+0x14/0x48
[ 115.055267] [<ffff00000882b89c>] ptp_cc_read+0x2c/0xe8
[ 115.075029] [<ffff00000813c79c>] timecounter_init+0x2c/0x58
[ 115.094949] [<ffff00000882babc>] otx2_ptp_init+0x164/0x240
[ 115.115228] [<ffff00000882654c>] otx2_probe+0x57c/0x6c8
[ 115.135140] [<ffff000008514744>] pci_device_probe+0xdc/0x170
[ 115.156475] [<ffff000008601b64>] driver_probe_device+0x1f4/0x298
[ 115.178178] [<ffff000008601cc4>] __driver_attach+0xbc/0xc0
[ 115.198386] [<ffff0000085ffd74>] bus_for_each_dev+0x4c/0x98
...

This patch fixes the problem.

Change-Id: Ifd8c59538ae71f3b96339e4a7720f033b81cbcf0
Signed-off-by: Bhatta Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/1755
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/otx2_ptp.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/otx2_ptp.c b/drivers/net/ethernet/marvell/octeontx2/otx2_ptp.c
index 564e2febbb9d..f38075443460 100644
--- a/drivers/net/ethernet/marvell/octeontx2/otx2_ptp.c
+++ b/drivers/net/ethernet/marvell/octeontx2/otx2_ptp.c
@@ -158,12 +158,14 @@ int otx2_ptp_init(struct otx2_nic *pfvf)
 		return 0;
 	}
 
-	ptp_ptr = kmalloc(sizeof(*ptp_ptr), GFP_KERNEL);
+	ptp_ptr = kzalloc(sizeof(*ptp_ptr), GFP_KERNEL);
 	if (!ptp_ptr) {
 		err = -ENOMEM;
 		goto error;
 	}
 
+	ptp_ptr->nic = pfvf;
+
 	spin_lock_init(&ptp_ptr->spin_lock);
 
 	cc = &ptp_ptr->cycle_counter;
@@ -198,8 +200,6 @@ int otx2_ptp_init(struct otx2_nic *pfvf)
 		goto error;
 	}
 
-	ptp_ptr->nic = pfvf;
-
 	pfvf->ptp = ptp_ptr;
 
 error:
-- 
2.17.1

